app/application.c

```diff
@@ -821,6 +821,17 @@ void bc_radio_on_buffer(uint64_t *peer_device_address, uint8_t *buffer, size_t *
             _radio_pub_state(buffer[0] == RADIO_RELAY_0_SET ? RADIO_RELAY_0 : RADIO_RELAY_1, buffer[sizeof(uint64_t) + 1]);
             break;
         }
+        case RADIO_RELAY_0_PULSE_SET:
+        case RADIO_RELAY_1_PULSE_SET:
+        {
+            if (*length != (1 + sizeof(uint64_t) + 1 + 4))
+            {
+                return;
+            }
+            bc_tick_t duration = buffer[sizeof(uint64_t) + 2] | (buffer[sizeof(uint64_t) + 3] << 8) | (buffer[sizeof(uint64_t) + 4] << 16) | (buffer[sizeof(uint64_t) + 5] << 24);
```

Comment:
Why not a simple memcpy instead of this?

Reply:
memcpy will do, thanks.
Is there a reason why we dont create C structures with pack attribute? I know that by someone this is not clean C solution, but this code will always be compiled to ARM Cortex core and adding a support for another compiler is easy by defining some macros. I use this approach on many projects and its the most elegant solution.
``` 
struct  __attribute__((__packed__)) my_struct {
    uint8_t command;
    uint64_t address;
    uint8_t direction;
    uint32_t duration;
};

my_struct packet;

packet.address = 12345
pacekt.command = xxx
packet.direction = direction
packet.duration = duration

bc_radio_pub_buffer(&packet, sizeof(packet));
```
This way we can get rid of ugly buffer[index + something], the adding of new parameters to the existing command is also easy.

Your opinions?

Reply:
I would say, the problem is way more complex than that: http://www.catb.org/esr/structure-packing/

In short, yes you can do that, but beware you may run across troubles when porting the code between different environments - which you almost always ultimately do (node x gateway). Thanks to re-ordering, padding, etc. you will sooner or later come across an issue even if using the same toolchain.

It is more portable to use:

```
#pragma pack(push)
#pragma pack(1)
typedef struct{
  char a;
  int b;
  char c;
}st_a;
#pragma pack(pop)
```

I have come across issue when using the same code on ARM GCC and MingGW GCC - the structures had different layouts on with `__attribute__((__packed__))` but the same layout with the example above. It was well-documented behaviour - I spent good amount of minutes on StackOverflow figuring out the right way :-)

I would say the situation is not black and white. We may go for struct-based syntax, but we should be really cautious about the real struct partitioning. Thx!